src/messages/program/publish.ts

```diff
@@ -0,0 +1,105 @@
+import { Account } from "../../accounts/account";
+import { ItemType, MessageType, ProgramMessage } from "../message";
+import { Publish as storePublish } from "../../messages/store/index";
+import { Encoding, FunctionTriggers, MachineType, MachineVolume, ProgramContent } from "./programModel";
+import { PutContentToStorageEngine } from "../create/publish";
+import { SignAndBroadcast } from "../create/signature";
+
+type ProgramPublishConfiguration = {
+    account: Account;
+    channel: string;
+    storageEngine: ItemType;
+    inlineRequested: boolean;
+    APIServer: string;
+    file: Buffer | Blob;
+    entrypoint: string;
+    subscription?: Record<string, unknown>[];
+    memory?: number;
+    runtime?: string;
+    volumes?: MachineVolume[];
+};
+
+// TODO: Check that program_ref, runtime and data_ref exist
+// Guard some numbers values
+export async function publish(configuration: ProgramPublishConfiguration): Promise<ProgramMessage> {
+    const timestamp = Date.now() / 1000;
+    const storageEngine: ItemType = ItemType.storage;
+
+    // Store the source code of the program and retrieve the hash.
+    const programRef = await storePublish({
+        channel: configuration.channel,
+        APIServer: configuration.APIServer,
+        account: configuration.account,
+        storageEngine: storageEngine,
+        fileObject: configuration.file,
+    });
+
+    let triggers: FunctionTriggers = { http: true };
+    if (configuration.subscription) {
+        triggers = { ...triggers, message: configuration.subscription };
+    }
+
+    const content: ProgramContent = {
+        address: configuration.account.address,
+        time: timestamp,
+        type: MachineType.vm_function,
+        allow_amend: false,
+        code: {
+            encoding: Encoding.zip, // retrieve the file format or params
+            entrypoint: configuration.entrypoint,
+            ref: programRef.item_hash,
+            use_latest: true,
+        },
+        on: triggers,
+        environment: {
+            reproducible: false,
+            internet: true,
+            aleph_api: true,
+            shared_cache: false,
+        },
+        resources: {
+            vcpus: 1,
+            memory: configuration.memory ? configuration.memory : 128,
+            seconds: 30,
+        },
+        runtime: {
+            ref: configuration.runtime
+                ? configuration.runtime
+                : "bd79839bf96e595a06da5ac0b6ba51dea6f7e2591bb913deccded04d831d29f4",
```

Comment:
What does this ref. is refering to ?

Reply:
runtime refers to the `root filesystem` used by the Aleph VM
and the ref param is a sort of path to them, you can find the official Aleph runtimes [here](https://github.com/aleph-im/aleph-vm/tree/main/runtimes)